browser-hist.el

```diff
@@ -66,108 +80,125 @@
+                   (mapconcat #'identity queries " AND ") rest
+                   (and emptyp " LIMIT 100")))) ;No match, just return history
+        (db (or browser-hist--db-connection
+                (setq browser-hist--db-connection
+                      (browser-hist--sqlite-open
+                       (browser-hist--db-copy-name
+                        browser-hist-default-browser))))))
+    (cl-loop for (desc link) in (browser-hist--sqlite-select db full-query)
+             unless (or (null desc) (string-blank-p desc))
+             collect (cons (string-trim-right
+                            (replace-regexp-in-string
+                             (if browser-hist-ignore-query-params "\\?.*" "")
+                             "" link))
+                           desc))))
+
+(defun browser-hist--completion-table (s _ flag)
+  (let* ((rows-raw (if (>= (length (string-trim s))
+                           browser-hist-minimum-query-length)
+                       (browser-hist--send-query s)
+                     (browser-hist--send-query nil))))
     (pcase flag
       ('metadata
        `(metadata
          (annotation-function
           ,@(lambda (x)
-              (concat
-               "\n\t"
-               (propertize
-                (alist-get x coll nil nil #'string=)
-                'face 'completions-annotations))))
-         (display-sort-function ; keep rows sorted as they come from db
-          ,@(lambda (xs) xs))
+              (concat " "
+               (if (> (length x) (floor (window-width) 2))
+                   "\n\t" (propertize " " 'display '(space :align-to (- center 1))))
+               (alist-get x rows-raw nil nil #'string=))))
+         (display-sort-function ,@(lambda (xs) xs))
          (category . url)))
-      ('t
-       (all-completions s coll)))))
+      ('nil (try-completion s rows-raw))
+      ('t (mapcar #'car rows-raw)))))
 
 (defun browser-hist--url-transformer (type target)
   "Remove title from TARGET url appended by `browser-hist-search'"
-  `(,type .
-    ,(replace-regexp-in-string "\t.*" "" target)))
+  `(,type . ,(replace-regexp-in-string "\t.*" "" target)))
 
 (defun browser-hist--url-handler (url &rest _)
   "Remove title from TARGET url appended by `browser-hist-search'"
   (browse-url (replace-regexp-in-string "\t.*" "" url)))
 
-(defun browser-hist-search ()
-  "Search through browser history."
-  (interactive)
+;;;###autoload
+(defun browser-hist-search (&optional force-update)
+  "Search through browser history.
+
+With prefix-arg FORCE-UPDATE, ensure that the history cache is
+updated."
+  (interactive "P")
   (unless (member '(".*\t" . browser-hist--url-handler)
                   browse-url-handlers)
     (add-to-list 'browse-url-handlers '(".*\t" . browser-hist--url-handler)))
 
   (when (boundp 'embark-transformer-alist)
-    (unless (member '(url . browser-hist--url-transformer)
+    (unless (memq '(url . browser-hist--url-transformer)
```

Comment:
```suggestion
       (let ((tfa '(url . browser-hist--url-transformer)))
           (unless (memq tfa embark-transformer-alist)
```
I noticed that byte-compile throws warning: 'memq' called with literal list...

Reply:
We could just go back to `member` then? `memq` is technically faster but `embark-transformer-alist` is tiny so it won't matter.

Reply:
Whichever you say is the best.

Reply:
I changed it back to `member`.